src/Core/ContainerInterface.php

```diff
@@ -0,0 +1,55 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.2.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Core;
+
+use League\Container\Definition\DefinitionInterface;
+use Psr\Container\ContainerInterface as PsrInterface;
+
+/**
+ * Interface for the Dependency Injection Container in CakePHP applications
+ *
+ * This interface extends the PSR-11 container interface and adds
+ * methods to add services and service providers to the container.
+ *
+ * The methods defined in this interface use the conventions provided
+ * by league/container as that is the library that CakePHP uses.
```

Comment:
Is this meant to be swappable with other container libraries at any point? The interface below exposes explicit league types.

Reply:
>Is this meant to be swappable with other container libraries at any point?  

Ideally yes. I don't think we'd swap libraries on users in a bugfix release but it could happen in a feature release.

> The interface below exposes explicit league types.

Yes, to define constructor arguments for a service you need to chain off of `add()` or use `extend()` and then chain method calls of DefinitionInterface. I don't have a good solution for these problems as adding additional interfaces for Definitions and ServiceProviders means we need to do more wrapping and sub-classing of league/container.

An alternative approach would be to skip this interface all together and only use League\Container. However, there are two problems that I was hoping to solve with having our own container interface:

1. Insulating our users from the potential future scenario where league/container becomes unmaintained. In this scenario we could either try to help with maintaining the library or forking and adding backwards compatibility aliases like we did with Chronos. This problem might not ever happen to and we could be over-engineering in the present to solve a problem that never happens.
2. Having our own interface allows other DI containers to be used if they implement the interfaces via shims. Plugins could potentially provide 'cake compatible' implementations of php-di or symfony/container if they wanted. This problem is the harder one to solve right now, and the current code doesn't deliver on it.

We could solve problem 2 by changing how `add()` is defined and make it look like:

```php
function add($id, $concrete = null, ?bool shared = null, array $definition = [])
```

Providing all the definition options as an array would give us better encapsulation but we'd sacrifice typechecking on all the definition steps which is not great.

Another solution would be to wrap the league/container definition with a proxy that implements our interface. We can use a subclass as easily because of how [league/container works]https://github.com/thephpleague/container/blob/master/src/Container.php#L85). Problem 2 could also be solved if those compatibility shims were written against the interfaces in league/container as well.

Reply:
I don't think it's necessarily bad that the interface exposes library types. 

If the container-specific types are going to be passed into external libraries, then there isn't much point in hiding them because if we do switch, then the entire ecosystem has to switch with us.

The alternative is providing 2 sets of the same interface and selecting which one to use at runtime.

Reply:
There are potential solutions that wouldn't result in us passing external library interfaces into plugins/applications but they require more work on our part. I'm not sure that work is justifiable given that we can either contribute upstream to help with any maintenance of league/container or fork and fix should the upstream maintainers disappear entirely (which I don't think will happen).

Exposing league/container types means that any adapter for another container library will need to partly implement interfaces from CakePHP and partly from league. I think alternate implementations will be very rare and that most folks will just stick with the defaults, making this entire thought exercise moot.